cmd/apply-crds/main.go

```diff
@@ -0,0 +1,182 @@
+	flag.Parse()
+
+	if len(crdsDir) == 0 {
+		log.Fatalf("CRDs directory is required")
+	}
+
+	for _, crdDir := range crdsDir {
+		if _, err := os.Stat(crdDir); os.IsNotExist(err) {
+			log.Fatalf("CRDs directory %s does not exist", crdsDir)
+		}
+	}
+}
+
+func main() {
+	Run()
+}
+
+func Run() {
+	ctx := context.Background()
+
+	initFlags()
+
+	config, err := ctrl.GetConfig()
+	if err != nil {
+		log.Fatalf("Failed to get Kubernetes config: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("Failed to create API extensions client: %v", err)
+	}
+
+	if err := walkCrdsDir(ctx, client); err != nil {
+		log.Fatalf("Failed to apply CRDs: %v", err)
+	}
+}
+
+// walkCrdsDir walks the CRDs directory and applies each YAML file.
+func walkCrdsDir(ctx context.Context, client *clientset.Clientset) error {
+	for _, crdDir := range crdsDir {
+		// Walk the directory recursively and apply each YAML file.
+		err := filepath.Walk(crdDir, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() || filepath.Ext(path) != ".yaml" {
+				return nil
+			}
+
+			log.Printf("Apply CRDs from file: %s", path)
+			if err := applyCRDsFromFile(ctx, client, path); err != nil {
+				return fmt.Errorf("apply CRD %s: %v", path, err)
+			}
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("walk the path %s: %v", crdsDir, err)
+		}
+	}
+	return nil
+}
+
+// applyCRDsFromFile reads a YAML file, splits it into documents, and applies each CRD to the cluster.
+func applyCRDsFromFile(ctx context.Context, client *clientset.Clientset, filePath string) error {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return fmt.Errorf("open file %q: %v", filePath, err)
+	}
+	defer file.Close()
+
+	// Create a decoder that reads multiple YAML documents.
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 4096)
+	var crdsToApply []*apiextensionsv1.CustomResourceDefinition
+	for {
+		crd := &apiextensionsv1.CustomResourceDefinition{}
+		if err := decoder.Decode(crd); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("decode YAML: %v", err)
+		}
+		if crd.GetObjectKind().GroupVersionKind().Kind != "CustomResourceDefinition" {
+			log.Printf("Skipping non-CRD object %s", crd.GetName())
+			continue
+		}
+		crdsToApply = append(crdsToApply, crd)
+	}
+
+	// Apply each CRD separately.
+	for _, crd := range crdsToApply {
+		err := wait.ExponentialBackoffWithContext(ctx, retry.DefaultBackoff, func(context.Context) (bool, error) {
+			if err := applyCRD(ctx, client, crd); err != nil {
+				//nolint:nilerr
+				return false, nil
+			}
+			return true, nil
+		})
+		if err != nil {
+			return fmt.Errorf("apply CRD %s: %v", crd.Name, err)
+		}
+	}
+	return nil
+}
+
+// applyCRD creates or updates the CRD.
+func applyCRD(ctx context.Context, client *clientset.Clientset, crd *apiextensionsv1.CustomResourceDefinition) error {
```

Comment:
I see that `client` only being used to instantiate the crdClient. Instead of passing `client` so far down the call stack, why not just init `crdClient` much earlier and pass that down the stack instead?

Reply:
IMO it is no big difference if we pass the pointer of the client to the functions or pass just the CRD interface, but I changed it in https://github.com/NVIDIA/k8s-operator-libs/pull/50/commits/b36bdfc7b555e96670f4cd47d61dd849c0479f2e.
The downside we have now is that we are not able to use something different than the CRD client. This should be okay for this tool anyway as it is mainly related to CRDs.